### PR TITLE
Avoid unchecked casts in net parser

### PR DESCRIPTION
**Pull request by tamird**
Once this and #77426 are in, I'll send another PR adding scope id parsing.

r? @dtolnay

### CODE REVIEW

File: library/std/src/net/parser.rs [hunk 1]
```diff
@@ -84,29 +107,20 @@ impl<'a> Parser<'a> {
...
-            while let Some(digit) = p.read_digit(radix) {
-                result = (result * radix) + digit;
-                digit_count += 1;
-                if digit_count > max_digits || result >= upto {
-                    return None;
-                }
+            let mut result = T::ZERO;
+            let mut parsed = false;
+
+            while let Some(digit) = p.read_atomically(|p| p.read_char()?.to_digit(radix)) {
```
Comment (pickfire) [inline library/std/src/net/parser.rs:117]: Do we need a second `read_atomically`?

Comment (tamird) [reply library/std/src/net/parser.rs:117]: Yes, because if the next thing we read is not a digit, we want to rewind the parser.

**Review by dtolnay — APPROVED** [review]
Thank you, this is terrific.

**Pull request merged**